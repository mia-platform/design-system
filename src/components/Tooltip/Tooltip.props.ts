/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { ReactNode } from 'react'

import { Placement, TriggerMode } from './Tooltip.types'

export type TooltipProps = {

    /**
     * The children nodes to be rendered as tooltip target.
     */
    children: ReactNode,

    /**
     * The DOM container of the tip, the default behavior is to create a div element in body.
     */
    getPopupContainer?: (triggerNode: HTMLElement) => HTMLElement

    /**
     * Defines the tooltip position relative to the target. Either:
     *
     * - "top" (default)
     * - "left"
     * - "right"
     * - "bottom"
     * - "topLeft"
     * - "topRight"
     * - "bottomLeft"
     * - "bottomRight"
     * - "leftTop"
     * - "leftBottom"
     * - "rightTop"
     * - "rightBottom"
     */
    placement?: Placement

    /**
     * The text shown in the tooltip.
     */
    title: ReactNode | (() => ReactNode)

    /**
     * Defines a single tooltip trigger mode or multiple ones by passing an array. The possible values are:
     *
     * - "hover" (default)
     * - "focus"
     * - "click"
     * - "contextMenu"
     */
    trigger?: TriggerMode | TriggerMode[],

    /**
     * Callback executed when visibility of the tooltip card is changed.
     */
    onOpenChange?: (open: boolean) => void,
}
