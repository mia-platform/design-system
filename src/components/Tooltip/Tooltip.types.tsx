export enum Placement {
    Top = 'top',
    Left = 'left',
    Right = 'right',
    Bottom = 'bottom',
    TopLeft = 'topLeft',
    TopRight = 'topRight',
    BottomLeft = 'bottomLeft',
    BottomRight = 'bottomRight',
    LeftTop = 'leftTop',
    LeftBottom = 'leftBottom',
    RightTop = 'rightTop',
    RightBottom = 'rightBottom'
}

export enum TriggerMode {
    Hover = 'hover',
    Focus = 'focus',
    Click = 'click',
    ContextMenu = 'contextMenu'
}
