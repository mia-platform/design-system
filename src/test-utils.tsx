/**
 * Copyright 2023 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { RenderOptions, RenderResult, render } from '@testing-library/react'
import { ConfigProvider } from 'antd'
import { ReactNode } from 'react'

import { ThemeContext, defaultTheme } from './components/ThemeProvider'
import { ANT_PREFIX } from './components/ThemeProvider/Ant'

/**
 * Wraps a React node with a series of providers for testing purposes.
 *
 * @remarks It adds the default theme to components
 * @remarks It adds the default Ant class prefix
 * @remarks It removes the css-do-not-override classes from Ant Design components
 *
 * @param {ReactNode} children - The React node to be wrapped.
 * @returns {ReactNode} The wrapped React node.
 */
const TestWrapper = ({ children }: {children: ReactNode}): ReactNode => (
  <ThemeContext.Provider value={defaultTheme}>
    <ConfigProvider prefixCls={ANT_PREFIX} theme={{ hashed: false }}>
      {children}
    </ConfigProvider>
  </ThemeContext.Provider>
)

/**
 * Custom render function for testing React components.
 *
 * @param {ReactNode} element - The React node to be rendered.
 * @param {RenderOptions} [options] - Options for rendering.
 * @returns {RenderResult} The result of rendering the React node.
 */
const customRender = (
  element: ReactNode,
  options?: Omit<RenderOptions, 'wrapper'>,
): RenderResult => render(element, { wrapper: TestWrapper, ...options })

export * from '@testing-library/react'
export {
  customRender as render,
  TestWrapper,
}
