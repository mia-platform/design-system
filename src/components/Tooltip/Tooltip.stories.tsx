/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Meta, StoryObj } from '@storybook/react'

import { Button } from '../Button'
import { Tooltip } from '.'
import { defaults } from './Tooltip'

const meta = {
  component: Tooltip,
  args: {
    ...defaults,
    title: 'Tooltip text',
    children: 'Hover me',
  },
  argTypes: {
    children: { control: false },
  },
} satisfies Meta<typeof Tooltip>

export default meta
type Story = StoryObj<typeof meta>

export const BasicExample: Story = {}

export const CustomPlacement: Story = {
  args: {
    ...meta.args,
    placement: Tooltip.Placement.BottomRight,
  },
}

export const TriggeredOnClick: Story = {
  args: {
    ...meta.args,
    trigger: Tooltip.TriggerMode.Click,
    children: <Button>{'Click me'}</Button>,
  },
}

export const HiddenWithEmptyTitle: Story = {
  args: {
    ...meta.args,
    title: '',
    children: 'Hover me, nothing will happen',
  },
}
