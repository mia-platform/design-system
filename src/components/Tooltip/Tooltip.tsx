/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { Tooltip as AntTooltip } from 'antd'
import { ReactElement } from 'react'

import { Placement, Trigger } from './Tooltip.types'
import { TooltipProps } from './Tooltip.props'

const arrowProp = { pointAtCenter: true }

export const defaults = {
  placement: Placement.Top,
  trigger: Trigger.Hover,
}

export const Tooltip = (
  {
    children,
    getPopupContainer,
    placement = defaults.placement,
    title,
    trigger = defaults.trigger,
    onOpenChange,
  } : TooltipProps
) : ReactElement => {
  return (
    <AntTooltip
      arrow={arrowProp}
      getPopupContainer={getPopupContainer}
      placement={placement}
      title={title}
      trigger={trigger}
      onOpenChange={onOpenChange}
    >
      {children}
    </AntTooltip>
  )
}

Tooltip.Placement = Placement
Tooltip.Trigger = Trigger
