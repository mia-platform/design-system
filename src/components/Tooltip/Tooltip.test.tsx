/**
 * Copyright 2024 Mia srl
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

import { act } from 'react'

import { RenderResult, render, screen, userEvent, waitFor } from '../../test-utils'
import { Tooltip } from './Tooltip'

const defaultProps = {
  title: 'Tooltip text',
  onOpenChange: jest.fn(),
  getPopupContainer: () => document.getElementById('popup-container') || document.body,
}

const renderTooltipWithWrapper = (props = defaultProps) : RenderResult => {
  return render(
    <div id="popup-container">
      <Tooltip {...props}>{'Tooltip trigger'}</Tooltip>
    </div>
  )
}

describe('Tooltip', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders on hover', async() => {
    const { asFragment } = renderTooltipWithWrapper()
    await act(async() => userEvent.hover(screen.getByText(/tooltip trigger/i)))
    await waitFor(() => expect(screen.getByText(/tooltip text/i)).toBeInTheDocument())

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders on click', async() => {
    const props = {
      ...defaultProps,
      trigger: Tooltip.TriggerMode.Click,
    }
    const { asFragment } = renderTooltipWithWrapper(props)

    await act(async() => userEvent.click(screen.getByText(/tooltip trigger/i)))
    await waitFor(() => expect(screen.getByText(/tooltip text/i)).toBeInTheDocument())

    expect(asFragment()).toMatchSnapshot()
  })

  it('renders correctly with custom placement', async() => {
    const props = {
      ...defaultProps,
      placement: Tooltip.Placement.RightBottom,
    }
    const { asFragment } = renderTooltipWithWrapper(props)

    await act(async() => userEvent.hover(screen.getByText(/tooltip trigger/i)))
    await waitFor(() => expect(screen.getByText(/tooltip text/i)).toBeInTheDocument())

    expect(asFragment()).toMatchSnapshot()
  })

  it('calls onOpenChange correctly', async() => {
    renderTooltipWithWrapper()
    await act(async() => userEvent.hover(screen.getByText(/tooltip trigger/i)))
    await waitFor(() => expect(defaultProps.onOpenChange).toHaveBeenCalledTimes(1))
    await act(async() => userEvent.unhover(screen.getByText(/tooltip trigger/i)))
    await waitFor(() => expect(defaultProps.onOpenChange).toHaveBeenCalledTimes(2))
  })

  it('does not render with empty title', async() => {
    const props = {
      ...defaultProps,
      title: '',
    }

    renderTooltipWithWrapper(props)
    await act(async() => userEvent.hover(screen.getByText(/tooltip trigger/i)))
    expect(screen.queryByText(/tooltip text/i)).not.toBeInTheDocument()
    expect(defaultProps.onOpenChange).not.toHaveBeenCalled()
  })
})
