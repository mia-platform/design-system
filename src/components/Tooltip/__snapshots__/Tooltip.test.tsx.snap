// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders correctly with custom placement 1`] = `
<DocumentFragment>
  <div
    id="popup-container"
  >
    <span
      class="mia-platform-tooltip-open"
    >
      Tooltip trigger
    </span>
    <div
      class="mia-platform-tooltip mia-platform-zoom-big-fast-appear mia-platform-zoom-big-fast-appear-prepare mia-platform-zoom-big-fast mia-platform-tooltip-placement-rightBottom"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <div
        class="mia-platform-tooltip-arrow"
        style="position: absolute;"
      />
      <div
        class="mia-platform-tooltip-content"
      >
        <div
          class="mia-platform-tooltip-inner"
          role="tooltip"
        >
          Tooltip text
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip renders on click 1`] = `
<DocumentFragment>
  <div
    id="popup-container"
  >
    <span
      class="mia-platform-tooltip-open"
    >
      Tooltip trigger
    </span>
    <div
      class="mia-platform-tooltip mia-platform-zoom-big-fast-appear mia-platform-zoom-big-fast-appear-prepare mia-platform-zoom-big-fast mia-platform-tooltip-placement-top"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <div
        class="mia-platform-tooltip-arrow"
        style="position: absolute; bottom: 0px; left: 0px;"
      />
      <div
        class="mia-platform-tooltip-content"
      >
        <div
          class="mia-platform-tooltip-inner"
          role="tooltip"
        >
          Tooltip text
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tooltip renders on hover 1`] = `
<DocumentFragment>
  <div
    id="popup-container"
  >
    <span
      class="mia-platform-tooltip-open"
    >
      Tooltip trigger
    </span>
    <div
      class="mia-platform-tooltip mia-platform-zoom-big-fast-appear mia-platform-zoom-big-fast-appear-prepare mia-platform-zoom-big-fast mia-platform-tooltip-placement-top"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <div
        class="mia-platform-tooltip-arrow"
        style="position: absolute; bottom: 0px; left: 0px;"
      />
      <div
        class="mia-platform-tooltip-content"
      >
        <div
          class="mia-platform-tooltip-inner"
          role="tooltip"
        >
          Tooltip text
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
